### DISCUSSION TITLE
Retrieve map JPEG for Deebot 900/901?

### DISCUSSION THREAD

**Discussion by jfdid** [mrbungle64/node-red-contrib-ecovacs-deebot#66] [Q&A] [answered]
Hi, great work on this project, I really appreciate it!
Can you tell me if the map retrieval works for the Deebot 900?  I'm using the example flow and it gives lots of info but "MapImage" doesn't seem to appear.
This is the output:
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "CurrentMapName", value: "unknown" }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "CurrentMapMID", value: "464441265" }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "CurrentMapIndex", value: 0 }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "Maps", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapVirtualBoundaries", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapSpotAreas", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapVirtualBoundaryInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapVirtualBoundaryInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapVirtualBoundaryInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapSpotAreaInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapVirtualBoundaryInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapSpotAreaInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapVirtualBoundaryInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapVirtualBoundaryInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapVirtualBoundaryInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapVirtualBoundaryInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapVirtualBoundaryInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapVirtualBoundaryInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapSpotAreaInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapSpotAreaInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapSpotAreaInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapSpotAreaInfo", value: object }
6/1/2022, 9:43:55 PM[node: a9af24533d1ac4f9](http://192.168.7.77:1880/#)
msg.payload : Object
{ type: "MapVirtualBoundaryInfo", value: object }

Thank you.

**Comment by mrbungle64** [comment]
@jfdid 

You can use the "Retrieve full map data object" node.
The `msg.payload` contains most of the card-related data incl. `mapImage.mapBase64PNG`.

But you need to have [canvas installed](https://github.com/mrbungle64/node-red-contrib-ecovacs-deebot#installation) this.

**Reply by mrbungle64** [reply]
@jfdid 

Did that help you?

**Reply by jfdid** [reply]
Hi Sascha, thank you very much for replying.  I ended up having a wheel motor issue with the robot - it's torn down on my work bench right now so I'll put the battery back in and test this soon.  Stay tuned :)

**Reply by jfdid** [reply]
Hi Sascha, I'm back up and running now.
I used the Retrieve Full map data object node to query the vacuum(with "include map image" set to true).  
Output to a debug node and I get a lot of msg.payloads but none seem to include mpaImage.mapBase64PNG.  Is it buried in one of teh objects or is it a head by itself?
I've had Canvas installed from the beginning using the instructions provided.
Is there a way to get you a log output on this?
Is the Deebot 900/901 confirmed to produce a map?
Thanks very much.

**Reply by mrbungle64** [reply]
> ...
> Is the Deebot 900/901 confirmed to produce a map?

Hi @jfdid 

Yes, it's confirmed. I justed checked that it's still working

Please make sure that you have checked "Include map image"/"Bild der Karte inkludieren"

**Comment by unclej84** [comment]
Hi the map is included in the object:

![image](https://user-images.githubusercontent.com/60842330/174456740-16682a66-f42a-4993-bc66-89d75c607e3c.png)

You can get it with `msg.payload.value[0].mapBase64PNG.split(';base64,').pop()`. Or you directly use the "Get map image as base64 data-URI (PNG format)" / "Bild der Karte abfragen (PNG-Format)" command in the command node to just get the image data instead.

**Comment by unclej84** [comment, marked as answer]
There is also a sample flow under import - examples - flows - node-red-contrib-ecovacs-deebot - Only laser unit models - save map image to file which shows this workflow
![image](https://user-images.githubusercontent.com/60842330/174457023-1678f342-3b7c-4f2b-b4d6-8c3d208dfc5d.png)

**Reply by jfdid** [reply]
Thanks unclej.  It doesn't look like it's coming through in that payload for me.  
Could the issue be Canvas isn't set up properly?  Or would it still come through on the payload, just not compile/save?

![image](https://user-images.githubusercontent.com/56010681/174457370-64652a6f-65f1-430f-9d6d-82439768b2db.png)

**Comment by unclej84** [comment]
Hi, I think as Sasha wrote before you need canvas installed in order to get the property in the payload.

**Comment by jfdid** [comment]
I appreciate your help on this!
I've tried that sample flow to no avail.  Using the "Retrieve map image data" command node does not return the MapDataObject above, just the other payloads.  If I use Return Full Map Data Object, I do get the MapDataObject but no image.
Firmware on the robot is 1.5.3.
I have Canvas installed but don't know how to debug to make sure it's working correctly.  Any tips here?  Or maybe a way to issue commands to the robot via the terminal in order to see an output/verbose log?

**Reply by mrbungle64** [reply]
@jfdid 
Had you already seen my reply?
https://github.com/mrbungle64/node-red-contrib-ecovacs-deebot/discussions/66#discussioncomment-2978258

**Comment by jfdid** [comment]
Hi.  I tried a fresh install on a fresh Rasbian image and it works great.  My issue was likely that I'm running node-red as an add-on for Home Assistant and is in a Docker container.  Will try a stand-alone installation.

You guys are amazing.  Thank you VERY much for creating this.  It's very helpful for our home automation with two of these vacuums.  

For anyone reading who's trying to get a JPEG from the base64 image, there's a "image-tools" node that can do the conversion to buffer and then output.
